zkevm-circuits/src/evm_circuit/execution/error_oog_call.rs

```diff
@@ -34,23 +43,37 @@ pub(crate) struct ErrorOOGCallGadget<F> {
 impl<F: Field> ExecutionGadget<F> for ErrorOOGCallGadget<F> {
     const NAME: &'static str = "ErrorOutOfGasCall";
 
-    const EXECUTION_STATE: ExecutionState = ExecutionState::ErrorOutOfGasCALL;
+    const EXECUTION_STATE: ExecutionState = ExecutionState::ErrorOutOfGasCall;
 
     fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
         let opcode = cb.query_cell();
         cb.opcode_lookup(opcode.expr(), 1.expr());
-        // TODO: add CallCode etc. when handle ErrorOutOfGasCALLCODE in furture
-        // implementation
+        let is_call = IsZeroGadget::construct(cb, opcode.expr() - OpcodeId::CALL.expr());
+        let is_callcode = IsZeroGadget::construct(cb, opcode.expr() - OpcodeId::CALLCODE.expr());
+        let is_delegatecall =
+            IsZeroGadget::construct(cb, opcode.expr() - OpcodeId::DELEGATECALL.expr());
+        let is_staticcall =
+            IsZeroGadget::construct(cb, opcode.expr() - OpcodeId::STATICCALL.expr());
+
+        // We do the responsible opcode check explicitly here because we're not
+        // using the SameContextGadget for CALL, CALLCODE, DELEGATECALL or
+        // STATICCALL.
         cb.require_equal(
-            "ErrorOutOfGasCall opcode is Call",
-            opcode.expr(),
-            OpcodeId::CALL.expr(),
+            "ErrorOutOfGasCall opcode should be CALL, CALLCODE, DELEGATECALL or STATICCALL",
+            is_call.expr() + is_callcode.expr() + is_delegatecall.expr() + is_staticcall.expr(),
+            1.expr(),
         );
 
         let rw_counter_end_of_reversion = cb.query_cell();
         let tx_id = cb.call_context(None, CallContextFieldTag::TxId);
         let is_static = cb.call_context(None, CallContextFieldTag::IsStatic);
-        let call_gadget = CommonCallGadget::construct(cb, 1.expr(), 0.expr(), 0.expr());
+
+        let call_gadget = CommonCallGadget::construct(
+            cb,
+            is_call.expr(),
+            is_callcode.expr(),
+            is_delegatecall.expr(),
+        );
```

Comment:
Just realized that we might need to also pass `is_static` to `CommonCallGadget` here, to make sure the `has_value` is `0` when it's `STATICCALL`, the changes in `CommonCallGadget` will be:

https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/e19ba01ecf343e7a331eae8d268070d84e89aeea/zkevm-circuits/src/evm_circuit/util/common_gadget.rs#L517-L521

```diff
     let has_value = select::expr( 
-        is_delegatecall.expr(), 
+        is_delegatecall.expr() + is_staticcall.expr(), 
         0.expr(), 
         1.expr() - value_is_zero.expr(), 
     ); 
```


Also in line 87 the second argument of `gas_cost_expr` should be passed with `is_call`.

Reply:
Make below fixes in 37e1e1e:

1. Add `is_staticcall` parameter to `CommonCallGadget::construct`.

2. Move `require_equal(is_call + is_callcode + is_delegatecall + is_staticcall, 1)` constraint from both `CallOpGadget` and `ErrorOOGCallGadget` into CommonCallGadget.

3. Fix `has_value = select(is_delegatecall,` to `has_value = select(is_delegatecall + is_staticcall,` in CommonCallGadget.

4. Fix to pass `is_call` (to replace `1.expr()`) into `call_gadget.gas_cost_expr` in ErrorOGGCallGadget.

Reply:
Looks much better now!